REDItools: python scripts for RNA editing detection by RNA-Seq data
===================================================================

Introduction
============
<p align-text="justify"> RNA editing is a post-transcriptional phenomenon
involving the insertion/deletion or substitution of specific bases in precise RNA localizations.
In human, RNA editing occurs by deamination of cytosine to uridine (C-to-U) or mostly by the
adenosine to inosine (A-to-I) conversion through ADAR enzymes. A-to-I substitutions may have
profound functional consequences and have been linked to a variety of human diseases including
neurological and neurodegenerative disorders or cancer. Next generation sequencing technologies
offer the unique opportunity to investigate in depth RNA editing even though no dedicated
software has been released up to now.

REDItools are simple python scripts conceived to facilitate the investigation of RNA editing
at large-scale and devoted to research groups that would to explore such phenomenon in own
data but donâ€™t have sufficient bioinformatics skills.
They work on main operating systems (although unix/linux-based OS are preferred), can handle reads from whatever
platform in the standard BAM format and implement a variety of filters.</p>

<a href="https://github.com/BioinfoUNIBA/REDItools/blob/master/README_1.md">REDItools V1 manual</a>
<br><br>
<a href="https://github.com/BioinfoUNIBA/REDItools2">REDItools V2 manual</a>
<p><b>Note. </b>REDItools V2 is the latest optimized, parallel multi-node version of REDItools.</p>
<p><b>Important. </b>Reditool_DNA_RNA.py v1.3 available at <a href="https://github.com/BioinfoUNIBA/REDItools/blob/master/NPscripts/REDItoolDnaRnav13.py" download>this link </a></p>

